tools/cryptogen/ca.go

```diff
@@ -0,0 +1,238 @@
+	KeyAlgorithm       string
+
+	// These fields are filled by the BuildCA() method.
+	Signer   crypto.Signer
+	SignCert *x509.Certificate
+}
+
+// SignCertParams describes the parameters for the SignCertificate() method.
+type SignCertParams struct {
+	OrgUnits       []string
+	AlternateNames []string
+	KeyUsage       x509.KeyUsage
+	ExtKeyUsage    []x509.ExtKeyUsage
+	PublicKey      crypto.PublicKey
+}
+
+type certParams struct {
+	Template   *x509.Certificate
+	Parent     *x509.Certificate
+	PublicKey  crypto.PublicKey
+	PrivateKey any
+}
+
+// BuildCA generates and saves the signing key pair in baseDir/name.
+func BuildCA(baseDir string, ca *CA) error {
+	err := os.MkdirAll(baseDir, 0o750)
+	if err != nil {
+		return errors.Wrapf(err, "cannot create directory %s", baseDir)
+	}
+
+	priv, err := GeneratePrivateKey(baseDir, ca.KeyAlgorithm)
+	if err != nil {
+		return err
+	}
+	ca.Signer = NewSignerFromPrivateKey(priv)
+
+	template := x509Template()
+	// this is a CA
+	template.IsCA = true
+	template.KeyUsage |= x509.KeyUsageDigitalSignature |
+		x509.KeyUsageKeyEncipherment | x509.KeyUsageCertSign |
+		x509.KeyUsageCRLSign
+	template.ExtKeyUsage = []x509.ExtKeyUsage{
+		x509.ExtKeyUsageClientAuth,
+		x509.ExtKeyUsageServerAuth,
+	}
+
+	// set the organization for the subject
+	subject := subjectTemplateAdditional(ca)
+	subject.Organization = []string{ca.Organization}
+	subject.CommonName = ca.Name
+
+	template.Subject = subject
+	template.SubjectKeyId, err = computeSKI(priv)
+	if err != nil {
+		return err
+	}
+
+	ca.SignCert, err = genCertificate(baseDir, ca.Name, certParams{
+		Template:   &template,
+		Parent:     &template,
+		PublicKey:  getPublicKey(priv),
+		PrivateKey: priv,
+	})
+	return err
+}
+
+// SignCertificate creates a signed certificate based on a built-in template and saves it in baseDir/name.
+func (ca *CA) SignCertificate(baseDir, name string, p SignCertParams) (*x509.Certificate, error) {
+	template := x509Template()
+	template.KeyUsage = p.KeyUsage
+	template.ExtKeyUsage = p.ExtKeyUsage
+
+	// set the organization for the subject
+	subject := subjectTemplateAdditional(ca)
+	subject.CommonName = name
+	subject.OrganizationalUnit = append(subject.OrganizationalUnit, p.OrgUnits...)
+
+	template.Subject = subject
+	for _, san := range p.AlternateNames {
+		// try to parse as an IP address first
+		ip := net.ParseIP(san)
+		if ip != nil {
+			template.IPAddresses = append(template.IPAddresses, ip)
+		} else {
+			template.DNSNames = append(template.DNSNames, san)
+		}
+	}
+
+	return genCertificate(baseDir, name, certParams{
+		Template:   &template,
+		Parent:     ca.SignCert,
+		PublicKey:  p.PublicKey,
+		PrivateKey: ca.Signer,
+	})
+}
+
+// computeSKI compute Subject Key Identifier using RFC 7093, Section 2, Method 4.
+func computeSKI(privKey crypto.PrivateKey) ([]byte, error) {
+	var raw []byte
+
+	// Marshall the public key
+	switch kk := privKey.(type) {
+	case *ecdsa.PrivateKey:
+		ecdhKey, err := kk.ECDH()
```

Comment:
SKI is supposed to be a hash of a public key. Here, ECDH() returns private key. I know this code is from the existing ca.go. However, the correct usage may be
```
		pubKey, err := kk.Public().(*ecdsa.PublicKey).ECDH()
		raw = pubKey.Bytes()
```
For ed25519, we already use the public key.